File: solutions/0092-reverse-linked-list-ii/solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reverseBetween(self, head, left, right):
        """
        :type head: Optional[ListNode]
        :type left: int
        :type right: int
        :rtype: Optional[ListNode]
        """
        h=head;
        l,a,h1,k=[],0,head,-1
        while(h!=None):
            a+=1;
            if a>=left and a<=right:
                l.append(h.val);
            h=h.next;
        a=0
        while(h1!=None):
            a+=1
            if a>=left and a<=right:
                h1.val=l[k]
                k-=1;
            
            h1=h1.next;
        return head;

File: solutions/0092-reverse-linked-list-ii/solution.cpp
// /**
//  * Definition for singly-linked list.
//  * struct ListNode {
//  * int val;
//  * ListNode *next;
//  * ListNode() : val(0), next(nullptr) {}
//  * ListNode(int x) : val(x), next(nullptr) {}
//  * ListNode(int x, ListNode *next) : val(x), next(next) {}
//  * };
//  */
// class Solution {
// public:
//     ListNode* reverseBetween(ListNode* head, int left, int right) {
//         if(!head || !head->next) return head;
        
//         ListNode *prev=NULL, *temp=NULL, *cur=head;
//         int a=0;

//         while(++a!=left)
//         {
//             prev=cur;
//             cur=cur->next;
//         }

//         ListNode *con = prev; 
//         ListNode *tail = cur; 
        
//         prev = NULL; 

//         while(left <= right--) 
//         {
//             temp=cur->next;
//             cur->next=prev;
//             prev=cur;
//             cur=temp;
//         }
//         if(con != NULL) {
//             con->next = prev;
//         } else {
//             head = prev; 
//         }
        
//         tail->next = cur; 
//         return head;
//     }
// };

/**

 * Definition for singly-linked list.

 * struct ListNode {

 *     int val;

 *     ListNode *next;

 *     ListNode() : val(0), next(nullptr) {}

 *     ListNode(int x) : val(x), next(nullptr) {}

 *     ListNode(int x, ListNode *next) : val(x), next(next) {}

 * };

 */

class Solution {

public:

    ListNode* reverseBetween(ListNode* head, int left, int right) {

        if(!head || !head->next) return head;

        ListNode *prev=NULL,*temp=NULL,*cur=head;

        int a=0;

        while(++a!=left)
        {
            prev=cur;
            cur=cur->next;
        }
        ListNode *tail=cur,*con=prev;prev=NULL;
        while(left<=right--)
        {
            temp=cur->next;
            cur->next=prev;
            prev=cur;
            cur=temp;
        }
        if(con != NULL)
            con->next = prev;
        else 
            head = prev;
        tail->next = cur;
        return head;
    }

};
